tx/textual/valuerenderer/duration.go

```diff
@@ -0,0 +1,207 @@
+	negative := normalizeDuration(duration)
+	factors := factorSeconds(duration.Seconds)
+	components := []string{}
+
+	if factors.days > 0 {
+		components = append(components, fmt.Sprintf("%d %s", factors.days, maybePlural("day", factors.days != 1)))
+	}
+	if factors.hours > 0 || len(components) > 0 && (factors.minutes > 0 || factors.seconds > 0 || duration.Nanos > 0) {
+		components = append(components, fmt.Sprintf("%d %s", factors.hours, maybePlural("hour", factors.hours != 1)))
+	}
+	if factors.minutes > 0 || len(components) > 0 && (factors.seconds > 0 || duration.Nanos > 0) {
+		components = append(components, fmt.Sprintf("%d %s", factors.minutes, maybePlural("minute", factors.minutes != 1)))
+	}
+	if factors.seconds > 0 || duration.Nanos > 0 {
+		components = append(components, formatSeconds(factors.seconds, duration.Nanos)+" "+maybePlural("second", factors.seconds != 1 || duration.Nanos > 0))
+	}
+
+	s := strings.Join(components, ", ")
+
+	if s == "" {
+		s = "0 seconds"
+	}
+
+	if negative {
+		s = "-" + s
+	}
+
+	w.Write([]byte(s))
+	return nil
+}
+
+var (
+	durRegexp = regexp.MustCompile(`^(-)?(?:([0-9]+) days?)?(?:, )?(?:([0-9]+) hours?)?(?:, )?(?:([0-9]+) minutes?)?(?:, )?(?:([0-9]+)(?:\.([0-9]+))? seconds?)?$`)
+)
+
+// Parse implements the ValueRenderer interface.
+func (dr durationValueRenderer) Parse(_ context.Context, r io.Reader) (protoreflect.Value, error) {
+	bz, err := io.ReadAll(r)
+	if err != nil {
+		return protoreflect.Value{}, err
+	}
+
+	parts := durRegexp.FindStringSubmatch(string(bz))
+	if parts == nil {
+		return protoreflect.Value{}, fmt.Errorf("bad duration format: %s", string(bz))
+	}
+
+	negative := parts[1] != ""
+	var days, hours, minutes, seconds, nanos int
+
+	if parts[2] != "" {
+		days, err = strconv.Atoi(parts[2])
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf(`bad number "%s": %w`, parts[2], err)
+		}
+	}
+	if parts[3] != "" {
+		hours, err = strconv.Atoi(parts[3])
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf(`bad number "%s": %w`, parts[3], err)
+		}
+	}
+	if parts[4] != "" {
+		minutes, err = strconv.Atoi(parts[4])
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf(`bad number "%s": %w`, parts[4], err)
+		}
+	}
+	if parts[5] != "" {
+		seconds, err = strconv.Atoi(parts[5])
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf(`bad number "%s": %w`, parts[5], err)
+		}
+		if parts[6] != "" {
+			if len(parts[6]) > 9 {
+				return protoreflect.Value{}, fmt.Errorf(`too many nanos "%s"`, parts[6])
+			}
+			addZeros := 9 - len(parts[6])
+			text := parts[6] + strings.Repeat("0", addZeros)
+			nanos, err = strconv.Atoi(text)
+			if err != nil {
+				return protoreflect.Value{}, fmt.Errorf(`bad number "%s": %w`, text, err)
+			}
+		}
+	}
+
+	dur := &dpb.Duration{}
+	dur.Seconds = int64(days)*int64(day_sec) + int64(hours*hour_sec+minutes*min_sec+seconds)
+	dur.Nanos = int32(nanos)
+
+	if negative {
+		dur.Seconds *= -1
+		dur.Nanos *= -1
+	}
+
```

Comment:
What about calling `dur.CheckValid` just to be sure our algo was correct?

Reply:
Outside the scope of `Parse()`. It doesn't need to handle anything outside of data formatted from a valid proto.